feat: add FAQ section (#271)
bytecodealliance/component-docs: component-model/src/reference/faq.md

# Frequently Asked Questions (FAQ)

This page hosts a series of questions that are frequently asked or that might be confusing about
WebAssembly (core), components, and the WebAssembly ecosystem as a whole.

## Q: What is the difference between a WebAssembly component and a WebAssembly module?

A WebAssembly module (more precisely referred to as a "WebAssembly core module") is a
binary that conforms to the [WebAssembly Core Specification][wasm-core-spec].

A WebAssembly component refers to a WebAssembly core module that has been wrapped in
the [Canonical ABI][cabi], i.e. the binary contract for the Component Model. By adhering to the
canonical ABI, WebAssembly components can take advantage of all the features the Component Model
has to offer.

WebAssembly Components can (and often do) contain core modules, but generally WebAssembly core modules
*cannot not* contain Components. One easy way to differentiate is by reading the WAT for a component:

> A WebAssembly core module generally starts with a `(module)` s-expression:

Author: Ah I see -- I didn't mean to imply that but this is a good point -- I'll change the wording here

```wat
(module

>   ...

Reviewer: Do we want to make etc. stuff as comments? I know the [MDN docs](https://developer.mozilla.org/en-US/docs/WebAssembly/Guides/Understanding_the_text_format) use this `...` notation but I'm not sure if it's _technically_ valid WAT.
```suggestion
  ;; ...
```

```

A WebAssembly component generally consists of a `(component)` s-expression (and may contain
nested `(module)`/`(component)` s-expressions):

```wat
(component
  ;; ...
)
```

> One part that might cause confusion here is that a WebAssembly Preview 1 "component" is in fact a

Author: ```suggestion
One part that might cause confusion here is that a WASI Preview 1 "component" is in fact a
```

*core module*. More precisely, a Preview 1 "component" is a WebAssembly core module with a well-defined
set of exports ([legacy specification][wasi-p1]).

[cabi]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md
[wasi-p1]: https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/witx/wasi_snapshot_preview1.witx
[wasm-core-spec]: https://webassembly.github.io/spec/core/

## Q: How do WebAssembly Components and the WebAssembly System Interface (WASI) relate to each other?

> WebAssembly core components do not have any access to the host system (by design) -- they can only perform

Author: ```suggestion
WebAssembly core modules do not have any access to the host system (by design) -- they can only perform
```

# Frequently Asked Questions (FAQ)

This page hosts a series of questions that are frequently asked or that might be confusing about
WebAssembly (core), components, and the WebAssembly ecosystem as a whole.

## Q: What is the difference between a _component_ and _module_ in WebAssembly?

A WebAssembly module (more precisely referred to as a "WebAssembly core module") is a
binary that conforms to the [WebAssembly Core Specification][wasm-core-spec].

A WebAssembly component refers to a WebAssembly core module that has been wrapped in

> the [Canonical ABI][cabi], i.e. the binary contract for the Component Model. By adhering to the

Reviewer 2: For my own understanding: is it _just_ that the module ABI conforms to CABI?  I thought there was more to it in terms of metadata etc. (like binary WITs getting embedded or something), and that CABI was just the runtime contract for passing non-primitive types across a module boundary?  Or is the metadata represented via CABI somehow?

Author: Ah good point -- the binary format on disk is different and it's better here to mention that!

- [ ] fix  high level description of components

Author: Reworded this, going to resolve this particular conversation but would appreciate another look on the next go-through! 🙇

canonical ABI, WebAssembly components can take advantage of all the features the Component Model
has to offer.

> }

Author: ```suggestion
```

WebAssembly Components can (and often do) contain core modules, but generally WebAssembly core modules

> *cannot not* contain Components. One easy way to differentiate is by reading the WAT for a component:

Author: Ah yeah, that's a great suggestion, will do!

- [ ] Split up questions about difference between component and module

A WebAssembly core module generally consists of a `(module)` s-expression:
```wat
(module
  ;; ...
)
```

A WebAssembly component generally consists of a `(component)` s-expression (and may contain
nested `(module)`/`(component)` s-expressions):

```wat
(component
  ;; ...
)
```

> One part that might cause confusion here is that a WebAssembly Preview 1 "component" is in fact a

Reviewer 2: I wonder if this belongs next to the "what is the difference between Preview 1 and Preview 2" question?

> One part that might cause confusion here is that a WebAssembly Preview 1 "component" is in fact a

Author: Ah thanks, yep that should be "WASI"

- [ ] WebAssembly -> WASI

*core module*. More precisely, a Preview 1 "component" is a WebAssembly core module with a well-defined

> set of exports ([specified in WITX][wasi-p1]).

Reviewer 2: I'd maybe drop the mention of WITX here.  It's a historical technology and we don't want people haring off in that direction unless they are specifically interested in P1 technologies.

Author: Yeah I could see that -- will reword this!
```suggestion
set of exports ([legacy specification][wasi-p1]).
```

> set of exports ([specified in WITX][wasi-p1]).

Author: Ah good catch, fixed!

[cabi]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md
[wasi-p1]: https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/witx/wasi_snapshot_preview1.witx
[wasm-core-spec]: https://webassembly.github.io/spec/core/

## Q: How do WebAssembly Components and the WebAssembly System Interface (WASI) relate to each other?

> WebAssembly core components do not have any access to the host system (by design) -- they can only perform

Author: Ah yes, so this should be "core module"

- [ ] "core component" -> "core module"

> computations on a fixed set of types. The component model enables core components to interact with the "outside"

Reviewer 2: This invites a question about "but core modules can affect the outside world via the DOM/browser" which I admit is neither true nor a question but hopefully you get what I mean

Author: I think "via a rich set of types" is the lynchpin here, I can try to make that more obvious with another sentence contrasting them

- [ ] Add sentence contrasting modules & component computations on types

> computations on a fixed set of types. The component model enables core components to interact with the "outside"

Author: I don't know that I would call the functions "pure", but maybe I can spell out the types here

- [ ] Spell out "fixed set of types"

> computations on a fixed set of types. The component model enables core components to interact with the "outside"

Author: ah, so core modules can only deal in types `i32`, `i64`, `f32`, `f64` and `v128` IIRC -- you can *represent* an arbitrary set of types via. those basic types, but the computation only deals in those, is what I was trying to get across.

The fact that those types don't match is part of why the CM exists in the first place ( to consistently handle higher level types) so that's the point I'm trying to get across

> computations on a fixed set of types. The component model enables core components to interact with the "outside"

Author: Ah I see -- yeah I had a bit of a think about this and reworded, what do you think of the new version?

> computations on a fixed set of types. The component model enables core components to interact with the "outside"

Reviewer 2: The second paragraph looks great.  I think you should just drop the first paragraph: for me, it sends a misleading message about the Wasm virtual machine in the component model.  And the second para stands alone and keeps the focus on "why and how is this tangled up with WASI."  What do you reckon?

world via a rich set of types ([WebAssembly Interface Types][wit]).

> The WebAssembly System Interface (WASI) is a *standardization* of the interfaces, functions and types that

Author: Thanks, I'll use the wording in the WASI repo

- [ ] Change to WASI repo wording over standard

a system or platform can expose to a WebAssembly component. At a glance, many parts of WASI are UNIX-like,
in that they match traditional expectations for programs like STDIN, STDOUT, and writing to files.

Some WASI system interfaces work at a much higher level than the command line however, like
[`wasi:http`][wasi-http]. `wasi:http` is included as a standardized platform due to the ubiquity
of the internet and the common use case of WebAssembly components with "the web" as a platform.

> With WIT, platform builders can define *any* interface to be *your* system interfaces that WebAssembly components can

Reviewer 2: > platform builders can define *any* interface to be *your* system interfaces that WebAssembly components can expect to access

Not sure what this is trying to say - did the sentence structure get muddled across multiple edits perhaps?  Is the idea that platform builders can implement the WASI abstractions in terms of the underlying platform?

expect to access -- WASI is a standardized set which enables to build on a shared base set of abstractions.

[wit]: https://component-model.bytecodealliance.org/design/wit.html
[wasi-http]: https://github.com/WebAssembly/wasi-http

## Q: I see the terms Preview 1 and Preview 2 frequently. What do those refer to?

Preview 1 refers to the first iteration of the Component Model which was based on WITX and is now deprecated:

https://github.com/WebAssembly/WASI/tree/main/legacy

Preview 2 refers to a newer iteration of the Component Model which uses WebAssembly Interface Types (WIT):

https://github.com/WebAssembly/WASI/tree/main/wasip2

Many programming language toolchains may only support Preview 1 components natively, but this isn't a problem
in practice as Preview 1 components can be *adapted* into Preview 2 components automatically.

## Q: What are component imports?

WebAssembly components represent a new kind of binary that can *describe* its interface or expected usage natively.

> This means that WebAssembly components have functionality that they might *import* (e.g. `wasi:cli/environment`).

Reviewer 2: "This means that" doesn't seem to follow.

I'd maybe rework this intro to something like:

> A component import declares that the component depends on an interface or function being provided to it at execution time - either by another component, or by the host runtime.

Author: I rewrote this -- still need to port changes to exports!

Imports are easiest illustrated with WIT:

```wit
package example-namespace:example-package;

world example-world {
    import wasi:cli/environment@0.2.4;
}
```

The [`environment` interface in `wasi:cli`][wasi-cli-env] provides various types and functions for interacting with
environment variables.

> The component is said to "import" the `wasi:cli/environment` interface, using the available functions and types therein.

Reviewer 2: I wonder if it would feel more concrete to write something like

> The component is said to "import" the `wasi:cli/environment` interface. Through this interface, the component will be able to interact with environment variables, using the functions and types that the interface provides.

(trying to communicate what the component is _doing_ with this mysterious interface, and that for imports it's the _component_ that consumes them)

Author: I think the link above covers the people who would be curious about what is in the interface.

The thing is, the import doesn't actually give the WebAssembly access to environment variables (it's not that direct)... so I didn't want to say it actually does.

[wasi-cli-env]: https://github.com/WebAssembly/wasi-cli/blob/main/wit/environment.wit

## Q: What are component exports?

> WebAssembly components represent a new kind of binary that can *describe* it's expected usage natively. This means that

Reviewer 2: ...or a COM typelib, or a .NET assembly, or a Java classfile... _grin_

> WebAssembly components represent a new kind of binary that can *describe* it's expected usage natively. This means that

Author: yeah, OK, maybe it's not *brand* new :)

WebAssembly components have functionality that they *export*, which users of the component (e.g. another component, or
a WebAssembly host) can use.

Exports are easiest illustrated with WIT:

```wit
package example-namespace:example-package;

interface example-interface {
    say-hello: func(name: string) -> string;
}

world example-world {
    export example-interface;
}
```

For the component that inhabits the `example-world` defined above, the outside world can expect the WebAssembly binary to
have a `say-hello` function that is callable via the `example-namespace:example-package/example-interface` interface.

> The component is said to "export" the `example-interface` interface, making available the functions and types therein.

Reviewer 2: Again I wonder if it would be useful to be more concrete, to help readers get a sense for the how the interface comes into play and who consumes the export.  What you're saying is pretty much implied by the previous paragraph, and it might be good to explore merging the two.

Author: Ah so this separation here is intention -- the repetition helps to enforce and sort of summarize the previous sentence, because of how dense it is.

WebAssembly Components can (and often do) contain core modules, but generally WebAssembly core modules
*cannot not* contain Components. One easy way to differentiate is by reading the WAT for a component:

A WebAssembly core module generally consists of a `(module)` s-expression:
```wat
(module
  ;; ...
)
```

A WebAssembly component generally consists of a `(component)` s-expression (and may contain
nested `(module)`/`(component)` s-expressions):

```wat
(component
  ;; ...
)
```

One part that might cause confusion here is that a WASI Preview 1 "component" is in fact a
*core module*. More precisely, a Preview 1 "component" is a WebAssembly core module with a well-defined
set of exports ([legacy specification][wasi-p1]).

[cabi]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md
[wasi-p1]: https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/witx/wasi_snapshot_preview1.witx
[wasm-core-spec]: https://webassembly.github.io/spec/core/

## Q: How do WebAssembly Components and the WebAssembly System Interface (WASI) relate to each other?

WebAssembly core modules do not have any access to the host system (by design) -- they can only perform

> computations on a fixed set of types. The component model enables core components to interact with the "outside"

Author: ```suggestion
computations on a fixed set of primitive types (`i32`, `i64`, `f32`, `f64`), `v128`). The component model enables core components to interact with the "outside"
```

> WebAssembly core modules do not have any access to the host system (by design) -- they can only perform
> computations on a fixed set of primitive types (`i32`, `i64`, `f32`, `f64`), `v128`). The component model enables core components to interact with the "outside"

Author: ```suggestion
WebAssembly core modules use functions provided by the host system (the "outside world") but they can only perform
computations on a fixed set of primitive types (`i32`, `i64`, `f32`, `f64`), `v128`). The component model enables core components to interact with the "outside world"
```

> world via a rich set of types ([WebAssembly Interface Types][wit]).

Author: ```suggestion
 via a rich set of types ([WebAssembly Interface Types][wit]).
```

> Where as WebAssembly core components must find a way to represent higher level types with the primitives defined in an ad hoc manner, the Component Model presents a consistent way of representing complex types familiar in most high level languages.

Author: ```suggestion
While WebAssembly core module *can* represent higher level types using the available primitives, every binary and platform may do so in an ad-hoc manner. The Component Model presents a consistent way of representing a rich set of types familiar in most high level languages that is consistent across binaries and platforms.
```

> The WebAssembly System Interface (WASI) is a *standardization* of the interfaces, functions and types that

Author: ```suggestion
The WebAssembly System Interface (WASI) is a set of APIs for WASI being developed for eventual standardization by the WASI Subgroup, which is a subgroup of the WebAssembly Community Group.

WASI defines interfaces, functions and types that
```

 via a rich set of types ([WebAssembly Interface Types][wit]).

While WebAssembly core module *can* represent higher level types using the available primitives, every binary and platform may do so in an ad-hoc manner. The Component Model presents a consistent way of representing a rich set of types familiar in most high level languages that is consistent across binaries and platforms.

The WebAssembly System Interface (WASI) is a set of APIs for WASI being developed for eventual standardization by the WASI Subgroup, which is a subgroup of the WebAssembly Community Group.

WASI defines interfaces, functions and types that
a system or platform can expose to a WebAssembly component. At a glance, many parts of WASI are UNIX-like,
in that they match traditional expectations for programs like STDIN, STDOUT, and writing to files.

Some WASI system interfaces work at a much higher level than the command line however, like
[`wasi:http`][wasi-http]. `wasi:http` is included as a standardized platform due to the ubiquity
of the internet and the common use case of WebAssembly components with "the web" as a platform.

> With WIT, platform builders can define *any* interface to be *your* system interfaces that WebAssembly components can

Author: ```suggestion
With WIT, platform builders can define *any* interface that WebAssembly components
```

expect to access -- WASI is a standardized set which enables to build on a shared base set of abstractions.

[wit]: https://component-model.bytecodealliance.org/design/wit.html
[wasi-http]: https://github.com/WebAssembly/wasi-http

## Q: I see the terms Preview 1 and Preview 2 frequently. What do those refer to?

Preview 1 refers to the first iteration of the Component Model which was based on WITX and is now deprecated:

https://github.com/WebAssembly/WASI/tree/main/legacy

Preview 2 refers to a newer iteration of the Component Model which uses WebAssembly Interface Types (WIT):

https://github.com/WebAssembly/WASI/tree/main/wasip2

Many programming language toolchains may only support Preview 1 components natively, but this isn't a problem
in practice as Preview 1 components can be *adapted* into Preview 2 components automatically.

## Q: What are component imports?

WebAssembly components are self-describing -- information about required external functionality (which must be provided by the platform or another component) is included in the binary.

> This means that WebAssembly components have functionality that they might *import* (e.g. `wasi:cli/environment`).

Author: ```suggestion
For example, a WebAssembly component that may require some use of outside environment variables may *import* a WASI interface like `wasi:cli/environment`.

> [!NOTE]
> The values provided by the `wasi:cli/environment` are not guaranteed
> to be ENV variables on the host machine -- this is a choice left to the
> platform, in the implementation of `wasi:cli/environment` that it exposes.
>
> For example, platforms may choose to elide sensitive environment variables, or provide none at all, in practice.
```

# Frequently Asked Questions (FAQ)

This page hosts a series of questions that are frequently asked or that might be confusing about
WebAssembly (core), components, and the WebAssembly ecosystem as a whole.

## Q: What is the difference between a _component_ and _module_ in WebAssembly?

A WebAssembly module (more precisely referred to as a "WebAssembly core module") is a
binary that conforms to the [WebAssembly Core Specification][wasm-core-spec].

> A WebAssembly component refers to a WebAssembly core module that has been wrapped in
> the [Canonical ABI][cabi], i.e. the binary contract for the Component Model. By adhering to the

Author: ```suggestion
A WebAssembly component is a WebAssembly binary that:
- Adheres to the component model [binary format][cm-binary-format] (as opposed to a WebAssembly core binary format)
- Uses the [WebAssembly Interface types][wit] specification to encode type information.
- Adheres to the Component Model [Canonical ABI][cabi] for converting between rich types and those present in core WebAssembly.
```

- Uses the [WebAssembly Interface types][wit] specification to encode type information.
- Adheres to the Component Model [Canonical ABI][cabi] for converting between rich types and those present in core WebAssembly.
WebAssembly Components can (and often do) contain core modules, but generally WebAssembly core modules
*cannot not* contain Components. One easy way to differentiate is by reading the WAT for a component:

A WebAssembly core module generally consists of a `(module)` s-expression:
```wat
(module
  ;; ...
)
```

A WebAssembly component generally consists of a `(component)` s-expression (and may contain
nested `(module)`/`(component)` s-expressions):

```wat
(component
  ;; ...
)
```

One part that might cause confusion here is that a WASI Preview 1 "component" is in fact a
*core module*. More precisely, a Preview 1 "component" is a WebAssembly core module with a well-defined

> set of exports ([legacy specification][wasi-p1]).

Author: ```suggestion
set of imports and exports ([legacy specification][wasi-p1]).
```

A WebAssembly component:
- Adheres to the component model [binary format][cm-binary-format] (as opposed to a WebAssembly core binary format)
- Uses the [WebAssembly Interface types][wit] specification to encode type information.
- Adheres to the Component Model [Canonical ABI][cabi] for converting between rich types and those present in core WebAssembly.

WebAssembly Components can (and often do) contain core modules, but generally WebAssembly core modules
*cannot* contain Components. WebAssembly components and WebAssembly core modules have a different binary format.

WebAssembly can be expressed via both a binary and textual format (["WAT", the WebAssembly Text format][wat]).

[wat]: https://webassembly.github.io/spec/core/text/index.html
[cabi]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md
[cm-binary-format]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/Binary.md
[wasi-p1]: https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/witx/wasi_snapshot_preview1.witx
[wasm-core-spec]: https://webassembly.github.io/spec/core/

## Q: How can I tell if a WebAssembly binary is a component or a module?

> After converting a WebAssembly binary to it's textual format (e.g. via a tool like [`wasm-tools print`][wasm-tools-examples]),

Author: Thanks for catching all these :sweat_smile:

https://github.com/WebAssembly/WASI/tree/main/wasip2

Many programming language toolchains may only support Preview 1 components natively, but this isn't a problem
in practice as Preview 1 components can be *adapted* into Preview 2 components automatically.

While somewhat confusing a WASI Preview 1 "component" is in fact a *WebAssembly core module*. More precisely, a
Preview 1 "component" is a WebAssembly core module with a well-defined set of imports and exports ([legacy specification][wasi-p1]).

## Q: What are component imports?

WebAssembly components are self-describing -- information about required external functionality (which must be provided by the platform or another component) is included in the binary.
For example, a WebAssembly component that may require some use of outside environment variables may *import* a WASI interface like `wasi:cli/environment`.

> [!NOTE]
> The values provided by the `wasi:cli/environment` are not guaranteed
> to be ENV variables on the host machine -- this is a choice left to the
> platform, in the implementation of `wasi:cli/environment` that it exposes.
>

> > For example, platforms may choose to elide sensitive environment variables, or provide none at all, in practice.

Author: Yeah, it just fell out after the refactor based on your comments, thanks again for that -- feels like we actually got a good concise explanation of a key benefit!
